sw/airborne/modules/optical_flow/px4flow.c

```diff
@@ -30,64 +31,134 @@
 #include "modules/datalink/mavlink_decoder.h"
 #include "subsystems/abi.h"
 
+// Messages
+#include "mcu_periph/uart.h"
+#include "pprzlink/messages.h"
+#include "subsystems/datalink/downlink.h"
+
+struct mavlink_heartbeat heartbeat;
 struct mavlink_optical_flow optical_flow;
-bool optical_flow_available;
+struct mavlink_optical_flow_rad optical_flow_rad;
 
 // message ID in Mavlink (v1.0)
+#define MAVLINK_OPTICAL_HEARTBEAT_MSG_ID 0
 #define MAVLINK_OPTICAL_FLOW_MSG_ID 100
-// size of the structure
-#define MAVLINK_OPTICAL_FLOW_LEN 26
+#define MAVLINK_OPTICAL_FLOW_RAD_MSG_ID 106
 
-#ifndef PX4FLOW_NOISE
-#define PX4FLOW_NOISE 0.5
+/*
+ * acceptable quality of optical flow (0-min,1-max)
+ * higher threshold means fewer measurements will be available
+ * but they will be more precise. Generally it is probably a better
+ * idea to have lower threshold, and propagate the noise (1-quality)
+ * to the INS
+ */
+#ifndef PX4FLOW_QUALITY_THRESHOLD
+#define PX4FLOW_QUALITY_THRESHOLD 0.1
 #endif
 
-// request struct for mavlink decoder
-struct mavlink_msg_req req;
+// request structs for mavlink decoder
+struct mavlink_msg_req req_flow;
+struct mavlink_msg_req req_flow_rad;
+struct mavlink_msg_req req_heartbeat;
 
 
 // callback function on message reception
 static void decode_optical_flow_msg(struct mavlink_message *msg __attribute__((unused)))
 {
-  optical_flow_available = true;
+  static float quality = 0;
+  static float noise = 0;
+  quality = ((float)optical_flow.quality)/255.0;
+  noise = 1-quality;
+
+  if (quality > PX4FLOW_QUALITY_THRESHOLD) {
+    // flip the axis (if the PX4FLOW is mounted as shown in
+    // https://pixhawk.org/modules/px4flow
+    AbiSendMsgVELOCITY_ESTIMATE(PX4FLOW_VELOCITY_ID,
+                                optical_flow.time_usec,
+                                optical_flow.flow_comp_m_y,
```

Comment:
It would be great, but then you would have to specify the orientation of the sensor etc. so for the time being the user will have to mount the sensor in a specific way, improvements are indeed welcome later.